onnxruntime/core/graph/function.cc

```diff
@@ -83,15 +83,72 @@ void IOTypeConstraintHelper(const ONNX_NAMESPACE::FunctionProto* onnx_func_proto
+// (2) The parent 'graph' - main model graph (OnnxRuntime::Graph)
+// (3) The node with a function body (ONNX::NodeProto)
+// (4) A map containing the input name from the op schema to the corresponding index
+// E.g. For Range-11, {"start" : 0, "limit": 1, "delta": 2}
+// (5) A map containing the output name from the op schema to the corresponding index
+// E.g. For Range-11, {"output" : 0}
+static void update_subgraph_nodes_within_function_nodes(ONNX_NAMESPACE::GraphProto& subgraph_proto,
+                                                        const Graph& parent_graph,
+                                                        const ONNX_NAMESPACE::NodeProto& function_node_in_parent_graph,
+                                                        const std::unordered_map<std::string, int>& input_name_idx_map,
+                                                        const std::unordered_map<std::string, int>& output_name_idx_map) {
+  // Iterate through all the nodes in the subgraph
+  for (const auto& subgraph_node : subgraph_proto.node()) {
+    // Iterate through all the inputs of the current node
+    for (int idx = 0; idx < subgraph_node.input_size(); ++idx) {
+      const auto& tensor_name = subgraph_node.input().Get(idx);
+      auto iter = input_name_idx_map.find(tensor_name);
+      // If an input pertaining to the name in the op schema is found,
+      // replace it with the corresponding input to the node with function body from the actual model graph
+      if (iter != input_name_idx_map.end()) {
+        const auto& parent_graph_input_to_function_node = function_node_in_parent_graph.input().Get(iter->second);
+        const_cast<ONNX_NAMESPACE::NodeProto&>(subgraph_node).set_input(idx, parent_graph_input_to_function_node);
+      }
+    }
+    // Iterate through all the output of the current node
+    for (int idx = 0; idx < subgraph_node.output_size(); ++idx) {
+      const auto& tensor_name = subgraph_node.output().Get(idx);
+      auto iter = output_name_idx_map.find(tensor_name);
+      if (iter != output_name_idx_map.end()) {
+        // If an input pertaining to the name in the op schema is found,
+        // replace it with the corresponding output to the node with function body from the actual model graph
+        const auto& parent_graph_output_to_function_node = function_node_in_parent_graph.output().Get(iter->second);
+        const_cast<ONNX_NAMESPACE::NodeProto&>(subgraph_node).set_output(idx, parent_graph_output_to_function_node);
+      }
+    }
+
+    for (auto& attr : subgraph_node.attribute()) {
+      if (attr.has_f()) {
+        ORT_THROW("A node with a function body within a subgraph within another function body is currently not supported in ORT");
+      }
+      // Recurse into any subgraphs in the current subgraph node being processed
+      if (attr.has_g()) {
+        update_subgraph_nodes_within_function_nodes(*const_cast<ONNX_NAMESPACE::AttributeProto&>(attr).mutable_g(),
```

Comment:
>*const_cast<ONNX_NAMESPACE::AttributeProto&>( [](start = 52, length = 45)

It would be more truthful to use non-const things in the 'for' loops given you are actually modifying the proto if you hit this (i.e. for (auto& attr : subgraph_node.mutable_attribute()) instead of const_cast (which should in general be avoided).

Reply:
It doesn't matter what we use in the for loop, I think. When we call attribute(), it returns back const stuff. I think the functionality (could be by design) is limited to get non-const mutable stuff back. Let me explore

Reply:
Fixed.